hazelcast/src/main/java/com/hazelcast/sql/impl/exec/scan/MapScanExecIterator.java

```diff
@@ -0,0 +1,135 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.exec.scan;
+
+import com.hazelcast.internal.serialization.Data;
+import com.hazelcast.internal.util.Clock;
+import com.hazelcast.map.impl.MapContainer;
+import com.hazelcast.map.impl.record.Record;
+import com.hazelcast.map.impl.recordstore.RecordStore;
+import com.hazelcast.sql.SqlErrorCode;
+import com.hazelcast.sql.impl.QueryException;
+
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * Iterator over map partitions..
+ */
+@SuppressWarnings("rawtypes")
+public class MapScanExecIterator {
+
+    private final MapContainer map;
+    private final Iterator<Integer> partsIterator;
+    private final long now = Clock.currentTimeMillis();
+
+    private RecordStore currentRecordStore;
+    private Iterator<Map.Entry<Data, Record<Object>>> currentRecordStoreIterator;
+
+    private Data currentKey;
+    private Object currentValue;
+    private Data nextKey;
+    private Object nextValue;
+
+    public MapScanExecIterator(MapContainer map, Iterator<Integer> partsIterator) {
+        this.map = map;
+        this.partsIterator = partsIterator;
+
+        advance0();
+    }
+
+    public boolean tryAdvance() {
+        if (hasNext()) {
+            currentKey = nextKey;
+            currentValue = nextValue;
+
+            advance0();
+
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    public boolean hasNext() {
+        return nextKey != null;
+    }
+
+    /**
+     * Get the next key/value pair from the store.
+     */
+    @SuppressWarnings("unchecked")
+    private void advance0() {
+        while (true) {
+            // Move to the next record store if needed.
+            if (currentRecordStoreIterator == null) {
+                if (!partsIterator.hasNext()) {
+                    nextKey = null;
+                    nextValue = null;
+
+                    return;
+                } else {
+                    int nextPart = partsIterator.next();
+
+                    boolean isOwned = map.getMapServiceContext().getOwnedPartitions().contains(nextPart);
+
+                    if (!isOwned) {
+                        throw QueryException.error(SqlErrorCode.PARTITION_MIGRATED,
+                            "Partition is not owned by member: " + nextPart);
+                    }
+
+                    currentRecordStore = map.getExistingRecordStore(nextPart);
+
+                    if (currentRecordStore == null) {
+                        // RecordStore might be missing if the associated partition is empty. Just skip it.
+                        continue;
+                    }
+
+                    currentRecordStore.checkIfLoaded();
```

Comment:
AFAIU, until all involved record stores are loaded, queries (or some parts of them?) would fail with `RetryableHazelcastException`. Not sure how `RetryableHazelcastException` is handled for SQL operations, but there are automatic retries for others. Looks like we need to check the possible "side effects" here.

Reply:
I'll take a look at it. Perhaps, a separate error code is needed here. Note that naive "retriable" semantics from the rest of HZ is not applicable here, because some data might already be returned to the user. So for the moment, I will only throw a special error code, and in the future we will have transparent retries when possible as explained in the PR description.

Reply:
Probably that's not a problem after all: in theory, all exceptions should be wrapped into `QueryException` and it's not marked as retryable.

Reply:
Yes, but here I lose the root cause - it is reported as GENERIC error. A separate error code is important here because this condition is potentially retriable. I'd better fix it now.

Reply:
Added separate error code.

Reply:
Just some afterthoughts: in theory, we can even "recover" from such situations since we still didn't send any data for the partition in question, so it's safe to return `IterationResult.WAIT`.

Reply:
This is dangerous - we do not know how much time the loading will take. So from the user standpoint, it may look like a hang.